internal/factsengine/factsengine.go

```diff
@@ -0,0 +1,174 @@
+package factsengine
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/trento-project/agent/internal/factsengine/adapters"
+	"github.com/trento-project/agent/internal/factsengine/gatherers"
+)
+
+type factsEngine struct {
+	agentID             string
+	factsEngineService  string
+	factGatherers       map[string]gatherers.FactGatherer
+	factsServiceAdapter adapters.Adapter
+}
+
+func NewFactsEngine(agentID, factsEngineService string) *factsEngine {
+	return &factsEngine{
+		agentID:            agentID,
+		factsEngineService: factsEngineService,
+		factGatherers: map[string]gatherers.FactGatherer{
+			gatherers.CorosyncFactKey: gatherers.NewCorosyncConfGatherer(),
+		},
+	}
+}
+
+func (c *factsEngine) Subscribe() error {
+	log.Infof("Subscribing agent %s to the facts gathering reception service on %s", c.agentID, c.factsEngineService)
+	//RabbitMQ adapter exists only by now
+	c.factsServiceAdapter = adapters.NewRabbitMQAdapter(c.factsEngineService)
+	log.Infof("Subscription to the facts engine by agent %s in %s done", c.agentID, c.factsEngineService)
+
+	return nil
+}
+
+func (c *factsEngine) Unsubscribe() error {
+	log.Infof("Unsubscribing agent %s from the facts engine service", c.agentID)
+	c.factsServiceAdapter.Unsubscribe()
```

Comment:
we should check the error here, and even in other defer clause similar to that, with a function and just log the error